### PR TITLE
Kubernetes experimentation

### PR DESCRIPTION
**Pull request by sriumcp**
Enables Kubernetes experimentation as described in #1157

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,45 +16,31 @@
...
+
+  > Grow your business with every release. Iter8 experiments can compare multiple versions based on business value and identify a winner.
+
+**Simple to use**
+
+  > Get started with Iter8 in seconds using pre-packaged experiment charts. Run Iter8 experiments locally, inside Kubernetes, or inside your CI/CD/GitOps pipelines.
+
+**App frameworks**
+
+  > Use with any app, serverless, or ML framework. Iter8 works with Kubernetes deployments, statefulsets, Knative services, KServe/Seldon ML deployments, or other custom Kubernetes resource types.
```
Comment (Alan-Cha) [inline README.md]:
```suggestion
  > Use with any app, serverless, or ML framework. Iter8 works with Kubernetes deployments, statefulSets, Knative services, KServe/Seldon ML deployments, or other custom Kubernetes resource types.
```

File: README.md [hunk 2]
```diff
@@ -16,45 +16,31 @@
...
 ## :dart: Usage examples
 1.  [Load test, benchmark and validate HTTP services with SLOs](https://iter8.tools/0.9/tutorials/load-test-http/usage/).
 2.  [Load test, benchmark and validate gRPC services with SLOs](https://iter8.tools/0.9/tutorials/load-test-grpc/usage/).
-3.  Load test, benchmark and validate Knative services with SLOs: [HTTP](https://iter8.tools/0.9/tutorials/integrations/knative/load-test-http/) and [gRPC](https://iter8.tools/0.9/tutorials/integrations/knative/load-test-grpc/).
+3.  Performance testing and SLO validation using Iter8 GitHub Action: [HTTP](http://localhost:8000/latest/tutorials/load-test-http/ghaction/) and [gRPC](http://localhost:8000/latest/tutorials/load-test-grpc/ghaction/)
+4.  Performance testing and SLO validation for Knative services: [HTTP](https://iter8.tools/0.9/tutorials/integrations/knative/load-test-http/) and [gRPC](https://iter8.tools/0.9/tutorials/integrations/knative/load-test-grpc/).
 
-### Documentation
-Iter8 documentation is available at https://iter8.tools.
+Please see https://iter8.tools for complete documentation.
```
Comment (Alan-Cha) [inline README.md]:
```suggestion
Please see [https://iter8.tools](https://iter8.tools) for the complete documentation.
```

File: base/experiment.go [hunk 3]
```diff
@@ -574,3 +583,219 @@ func (in *Insights) GetMetricsInfo(nm string) (*MetricMeta, error) {
...
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// NoFailure returns true if no task in the experiment has failed
+func (exp *Experiment) NoFailure() bool {
+	if exp != nil {
```
Comment (Alan-Cha) [inline base/experiment.go:611]:
```suggestion
return exp != nil && exp.Result != nil && !exp.Result.Failure
```

File: base/experiment.go [hunk 4]
```diff
@@ -574,3 +583,219 @@ func (in *Insights) GetMetricsInfo(nm string) (*MetricMeta, error) {
...
+		if exp.Result != nil {
+			if !exp.Result.Failure {
+				return true
+			}
+		}
+	}
+	return false
```
Comment (Alan-Cha) [inline base/experiment.go]:
```suggestion
	return exp != nil && exp.Result != nil && !exp.Result.Failure
```

**Review by Alan-Cha — DISMISSED** [review]

File: README.md (hunk 1 above)
Comment (sriumcp) [reply README.md]: Is the extra space in front of `>` intentional?

Comment (sriumcp) [reply README.md]: It seems like I can commit your suggestions to my PR directly through the UI without having to do it in my local repo... hence the odd question. More context [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request).

File: Dockerfile [hunk 5]
```diff
@@ -10,41 +10,43 @@ COPY ./ ./
...
-RUN curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | bash
-RUN cp kustomize /bin
+# # Install Kustomize v3
+# RUN curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | bash
+# RUN cp kustomize /bin
 
-# Install yq
-RUN GO111MODULE=on GOBIN=/bin go get github.com/mikefarah/yq/v4
+# # Install yq
+# RUN GO111MODULE=on GOBIN=/bin go get github.com/mikefarah/yq/v4
```
Comment (kalantar) [inline Dockerfile:35]: Why are we making these changes?  These were here to support the exec action.

Comment (sriumcp) [reply Dockerfile:35]: Good catch. As we introduce more capabilities in the charts, for example, notifications, we can reassess which of these binaries are actually required and put them back into the docker image.
***
For now, they are simply bloating the image, without actually contributing to the only two experiments we support through the charts today, namely `load-test-http` and `load-test-grpc`. Hence, the commenting out ...

**Comment by kalantar** [comment]
I notice that the `hub`,`launch` and `k launch` commands no longer support the `--chartVersionConstraint` and `--repoURL` options. I think at least the `--repoURL` needs to be supported.

**Comment by kalantar** [comment]
Out of curiosity, what are the `doc.go` files for?

**Comment by sriumcp** [comment]
> I notice that the hub,launch and k launch commands no longer support the --chartVersionConstraint and --repoURL options. I think at least the --repoURL needs to be supported.

They are all still supported. Just hidden. We can unhide the flags if need be. In fact, these flags are used for testing.

> Out of curiosity, what are the doc.go files for?

Unambiguous package level documentation for multi-file go packages. See [here](https://github.com/natefinch/godocgo/blob/master/sub/doc.go) for example and also [here](https://go.dev/doc/effective_go) (under commentary).

**Comment by kalantar** [comment]
>> I notice that the hub,launch and k launch commands no longer support the --chartVersionConstraint and --repoURL options. I think at least the --repoURL needs to be supported.

> They are all still supported. Just hidden. We can unhide the flags if need be. In fact, these flags are used for testing.

Ok. so I do see `--repoURL` defined in `NewHubOpts()`.  However, I do not see  `--chartVersionContraint` at all. Where is it?

**Comment by sriumcp** [comment]
> However, I do not see --chartVersionContraint at all. Where is it?

[Here](https://github.com/sriumcp/iter8/blob/17338758f79f285b04289f8ef1aabf1ae3669ce6/action/hub.go#L26). Also see [here](https://github.com/helm/helm/blob/1ec0aacb8865d5b1f7ef1cb884bbf9b12579ecef/pkg/action/install.go#L119) and [here](https://github.com/sriumcp/iter8/blob/17338758f79f285b04289f8ef1aabf1ae3669ce6/cmd/root.go#L75).

**Comment by sriumcp** [comment]
@kalantar Is this PR ready to merge or are there any comments / changes required?

**Review by kalantar — APPROVED** [review]
looks good.

**Pull request merged**